File: internal/watcher/watcher.go
package watcher

import (
	"context"
	"path/filepath"
	"sync"
	"time"

	"github.com/fsnotify/fsnotify"
	"github.com/sirupsen/logrus"
)

// SyncService interface for triggering syncs
type SyncService interface {
	SyncAllSCIM(ctx context.Context) error
}

// FileWatcher watches directories for changes and triggers sync operations
type FileWatcher struct {
	watcher     *fsnotify.Watcher
	syncService SyncService
	logger      *logrus.Logger
	debouncer   *Debouncer
	watchedDirs []string
}

// Debouncer prevents rapid successive syncs by batching events
type Debouncer struct {
	mu       sync.Mutex
	timer    *time.Timer
	callback func()
	delay    time.Duration
}

// NewDebouncer creates a new debouncer with the specified delay
func NewDebouncer(delay time.Duration, callback func()) *Debouncer {
	return &Debouncer{
		callback: callback,
		delay:    delay,
	}
}

// Trigger schedules the callback to run after the delay
// If called again before the delay expires, the timer is reset
func (d *Debouncer) Trigger() {
	d.mu.Lock()
	defer d.mu.Unlock()

	if d.timer != nil {
		d.timer.Stop()
	}

	d.timer = time.AfterFunc(d.delay, func() {
		d.callback()
	})
}

// NewFileWatcher creates a new file watcher
func NewFileWatcher(syncService SyncService, logger *logrus.Logger) (*FileWatcher, error) {
	watcher, err := fsnotify.NewWatcher()
	if err != nil {
		return nil, err
	}

	fw := &FileWatcher{
		watcher:     watcher,
		syncService: syncService,
		logger:      logger,
		watchedDirs: []string{},
	}

	// Create debouncer with 2 second delay to batch rapid changes
	fw.debouncer = NewDebouncer(2*time.Second, func() {
		fw.triggerSync()
	})

	return fw, nil
}

// AddDirectory adds a directory to watch for changes
func (fw *FileWatcher) AddDirectory(dir string) error {
	absPath, err := filepath.Abs(dir)
	if err != nil {
		return err
	}

	if err := fw.watcher.Add(absPath); err != nil {
		return err
	}

	fw.watchedDirs = append(fw.watchedDirs, absPath)
	fw.logger.WithField("directory", absPath).Info("Watching directory for changes")
	return nil
}

// Start begins watching for file system events
func (fw *FileWatcher) Start(ctx context.Context) {
	go func() {
		fw.logger.Info("File watcher started")
		defer fw.logger.Info("File watcher stopped")

		for {
			select {
			case event, ok := <-fw.watcher.Events:
				if !ok {
					return
				}

				// Only trigger on write, create, remove, rename events
				if event.Op&(fsnotify.Write|fsnotify.Create|fsnotify.Remove|fsnotify.Rename) != 0 {
					// Only process .json files (SCIM data files)
					if filepath.Ext(event.Name) == ".json" {
						fw.logger.WithFields(logrus.Fields{
							"file":      event.Name,
							"operation": event.Op.String(),
						}).Info("Detected file change")

						// Debounce to avoid multiple rapid syncs
						fw.debouncer.Trigger()
					}
				}

			case err, ok := <-fw.watcher.Errors:
				if !ok {
					return
				}
				fw.logger.WithError(err).Error("File watcher error")

			case <-ctx.Done():
				fw.logger.Info("File watcher context cancelled")
				return
			}
		}
	}()
}

// triggerSync performs the actual sync operation
func (fw *FileWatcher) triggerSync() {
	fw.logger.Info("Triggering SCIM-to-IAM sync due to file changes")

	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	if err := fw.syncService.SyncAllSCIM(ctx); err != nil {
		fw.logger.WithError(err).Error("Failed to sync SCIM data after file change")
	} else {
		fw.logger.Info("Successfully synced SCIM data after file change")
	}
}

// Close stops watching and cleans up resources
func (fw *FileWatcher) Close() error {
	if fw.debouncer.timer != nil {
		fw.debouncer.timer.Stop()
	}
	return fw.watcher.Close()
}

File: py_gateway/app/watcher.py
"""Debounced filesystem watcher.

Port of internal/watcher/watcher.go. Uses the ``watchdog`` library.

Watches one or more directories for ``.json`` file changes and calls the
configured sync callback after a quiet period (default 2 s), matching the
debouncer behaviour in the Go implementation.
"""

import logging
import threading
from pathlib import Path as FSPath
from typing import Callable, List, Optional

try:
    from watchdog.events import FileSystemEventHandler
    from watchdog.observers import Observer
except ImportError as _err:
    raise ImportError(
        "watchdog is required for file-change watching. "
        "Install it with: pip install 'watchdog>=3.0'"
    ) from _err

logger = logging.getLogger(__name__)


class _DebounceHandler(FileSystemEventHandler):
    """Debounce rapid filesystem events and call *callback* once they settle."""

    def __init__(self, callback: Callable, delay: float = 2.0) -> None:
        super().__init__()
        self._callback = callback
        self._delay = delay
        self._timer: Optional[threading.Timer] = None
        self._lock = threading.Lock()

    def _arm(self) -> None:
        """(Re)start the debounce timer."""
        with self._lock:
            if self._timer is not None:
                self._timer.cancel()
            self._timer = threading.Timer(self._delay, self._callback)
            self._timer.daemon = True
            self._timer.start()

    # Only react to .json files (SCIM data / policy changes).
    def on_created(self, event):
        if not event.is_directory and event.src_path.endswith(".json"):
            self._arm()

    def on_modified(self, event):
        if not event.is_directory and event.src_path.endswith(".json"):
            self._arm()

    def on_deleted(self, event):
        if not event.is_directory and event.src_path.endswith(".json"):
            self._arm()

    def on_moved(self, event):
        if not event.is_directory:
            self._arm()

    def cancel(self) -> None:
        with self._lock:
            if self._timer is not None:
                self._timer.cancel()
                self._timer = None


class FileWatcher:
    """Watch multiple directories and trigger *sync_callback* on JSON changes.

    Args:
        sync_callback:    Called (no arguments) after the debounce quiet period.
        health_refresher: Optional callback invoked after each triggered sync.
        debounce_delay:   Seconds of quiet time before the callback fires
                          (mirrors the 2 s Go debouncer).
    """

    def __init__(
        self,
        sync_callback: Callable,
        health_refresher: Optional[Callable] = None,
        debounce_delay: float = 2.0,
    ) -> None:
        self._sync_callback = sync_callback
        self._health_refresher = health_refresher
        self._observer = Observer()
        self._handler = _DebounceHandler(self._on_change, delay=debounce_delay)
        self._dirs: List[str] = []

    def _on_change(self) -> None:
        logger.info("File change detected – triggering SCIM → IAM sync")
        try:
            self._sync_callback()
        except Exception as exc:
            logger.error("Sync after file change failed: %s", exc)
        else:
            logger.info("Post file-change sync completed")
        if self._health_refresher:
            try:
                self._health_refresher()
            except Exception:
                pass

    def add_directory(self, path: str) -> None:
        """Watch *path* (non-recursive) for JSON file changes."""
        abs_path = str(FSPath(path).resolve())
        self._observer.schedule(self._handler, abs_path, recursive=False)
        self._dirs.append(abs_path)
        logger.info("Watching: %s", abs_path)

    def start(self) -> None:
        self._observer.start()
        logger.info("File watcher started (%d dir(s))", len(self._dirs))

    def close(self) -> None:
        self._handler.cancel()
        self._observer.stop()
        self._observer.join()
        logger.info("File watcher stopped")
